### DISCUSSION TITLE
How to write a class implementing `FromClause`?

### DISCUSSION THREAD

**Discussion by TommasoAmici** [sqlalchemy/sqlalchemy#9077] [Usage Questions] [answered]
Hello, I'm trying to create a `FromClause` to work with the `crosstab` function in Postgres ([docs](https://www.postgresql.org/docs/current/tablefunc.html#id-1.11.7.52.5.7)).

I found this https://github.com/makmanalp/sqlalchemy-crosstab-postgresql, which starts with a FromClause, but it's very outdated (2018, Python 2).

This discussion regarding adding PIVOT also suggests using a `FromClause`:

https://github.com/sqlalchemy/sqlalchemy/discussions/8029#discussioncomment-2767582

> the above could be turned into a slicker looking FromClause recipe if desired, where you could provide a `select()` and `func.avg(...)` expression as arguments along with the columns.

Is there any documentation on how I could go about doing this or at least an example in the source of SQLAlchemy I could look at?

I'm having trouble finding anything to get me started, and I'd like not to reverse engineer all of SQLAlchemy to get there :/

**Comment by CaselIT** [comment, marked as answer]
Hi,

I think you can just use `func.crosstab` with:
- use [table_valued](https://docs.sqlalchemy.org/en/14/core/functions.html#sqlalchemy.sql.functions.FunctionElement.table_valued)
- then call [render_derived](https://docs.sqlalchemy.org/en/14/core/selectable.html#sqlalchemy.sql.expression.TableValuedAlias.render_derived)

so the example in the pg docs would be something like this:
```py
ct = func.crosstab('the query sql string').table_valued(
 sa.column('row_name', sa.Text()), sa.column('extra', sa.Text()), sa.column('cat1', sa.Text()), # etc
).render_derived('ct',with_types=True)
```

(have not tried this, so it may require some changes)

**Reply by TommasoAmici** [reply]
This works like a charm, thanks a lot!

**Comment by zzzeek** [comment]
if the func approach above doesn't work, FromClause requires a bit of private API access that we've never made a real API for.   below is a brief adaptation of the ``values`` example at https://github.com/sqlalchemy/sqlalchemy/wiki/PGValues, adapted for 1.4:

```py
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.sql.expression import FromClause


class values(FromClause):
    named_with_column = True

    def __init__(self, columns, *args, **kw):
        super().__init__()
        self._column_args = columns
        self.list = args
        self.alias_name = self.name = kw.pop("alias_name", None)

    def _populate_column_collection(self):
        for c in self._column_args:
            self._columns.add(c)
            c.table = self

    @property
    def _from_objects(self):
        return [self]

```
